### PR TITLE
CLN: Use ABC classes where possible

### LINKED ISSUE #27353: Ensure isinstance checks use ABC classes where available
Here's a quick little script I wrote to check for things like isinstance(obj, MultiIndex)

```python
import ast
import pathlib


# Taken from pandas.core.dtypes.generic
BAD_NAMES = {'Index', 'Int64Index', 'UInt64Index', 'RangeIndex',
             'Float64Index', 'MultiIndex', 'DatetimeIndex', 'TimedeltaIndex',
             'PeriodIndex', 'CategoricalIndex', 'IntervalIndex', 'IndexClass',
             'Series', 'DataFrame', 'SparseDataFrame', 'SparseSeries',
             'SparseArray', 'Categorical', 'DatetimeArray', 'TimedeltaArray',
             'PeriodArray', 'Period', 'DateOffset', 'Interval', 'ExtensionArray',
             'PandasArray'}

def check_name_or_attr(filename, node):
    if isinstance(node, ast.Name):  # direct name
        if node.id in BAD_NAMES:
            print(f"{filename}:{node.lineno} isinstance check of {node.id}")
    elif isinstance(node, ast.Attribute):  # ex: pd.DataFrame
        if node.attr in BAD_NAMES:
            print(f"{filename}:{node.lineno} isinstance check of {node.attr}")
    

for path in pathlib.Path('.').glob('**/*.py'):
    filename = path.resolve()

    with open(path) as fh:
        contents = fh.read()
        tree = ast.parse(contents)

        for node in ast.walk(tree):
            try:
                isinstance_check = node.value.func.id == 'isinstance'
            except AttributeError:
                continue

            if isinstance_check:
                types = node.value.args[1]

                if isinstance(types, ast.Tuple):
                    for elt in types.elts:
                        check_name_or_attr(filename, elt)
                else:
                    check_name_or_attr(filename, types)

```

Running that from project root yields all of the following errors:

```sh
/Users/williamayd/clones/pandas/pandas/core/frame.py:4681 isinstance check of MultiIndex
/Users/williamayd/clones/pandas/pandas/core/indexing.py:2499 isinstance check of MultiIndex
/Users/williamayd/clones/pandas/pandas/core/indexing.py:564 isinstance check of MultiIndex
/Users/williamayd/clones/pandas/pandas/core/reshape/tile.py:540 isinstance check of Series
/Users/williamayd/clones/pandas/pandas/core/reshape/concat.py:373 isinstance check of DataFrame
/Users/williamayd/clones/pandas/pandas/core/reshape/concat.py:377 isinstance check of Series
/Users/williamayd/clones/pandas/pandas/core/computation/align.py:72 isinstance check of Series
/Users/williamayd/clones/pandas/pandas/core/arrays/period.py:931 isinstance check of Period
/Users/williamayd/clones/pandas/pandas/core/arrays/period.py:932 isinstance check of Period
/Users/williamayd/clones/pandas/pandas/core/indexes/base.py:3626 isinstance check of MultiIndex
/Users/williamayd/clones/pandas/pandas/core/indexes/base.py:3627 isinstance check of MultiIndex
/Users/williamayd/clones/pandas/pandas/io/formats/excel.py:453 isinstance check of Index
/Users/williamayd/clones/pandas/pandas/io/formats/excel.py:501 isinstance check of Index
/Users/williamayd/clones/pandas/pandas/io/formats/excel.py:552 isinstance check of Index
/Users/williamayd/clones/pandas/pandas/io/formats/excel.py:592 isinstance check of Index
```

So I think these need to be swapped out for their ABC equivalents

### PR DESCRIPTION
**Pull request by kgoehner**
- [x] closes #27353
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Looks like the linter doesn't like the import order.  can you run `isort pandas/core/reshape/concat.py`

**Comment by TomAugspurger** [comment]
This will need to be profiled carefully given @jorisvandenbossche's comments in https://github.com/pandas-dev/pandas/issues/27353#issuecomment-541301711.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
     ABCDataFrame,
     ABCDatetimeArray,
     ABCDatetimeIndex,
-    ABCIndexClass,
```
Comment (WillAyd) [inline pandas/core/indexes/base.py]: Why was this removed?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good generally. w.r.t. performance none of these look like they are in tight loop constructs so I think OK

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (Kazz47) [reply pandas/core/indexes/base.py]: Since `Index` implements `ABCIndexClass`, the check on line 516 will be faster if we just use `Index` itself. Since thats the only instance, we can then remove this. It's a small tradeoff I can put back if you would prefer to keep using ABCIndexClass here.

Basically my understanding is if something actually requires Index in the file we should check instances with Index directly, otherwise we should import ABCIndexClass to save on importing a larger file.

Comment (WillAyd) [reply pandas/core/indexes/base.py]: Hmm it's a good thought but I'd maybe leave that for a separate discussion to address more comprehensively

Comment (Kazz47) [reply pandas/core/indexes/base.py]: Alright, posed an update.

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -353,7 +353,7 @@ def __new__(
             return Index(data, dtype=object, copy=copy, name=name, **kwargs)
 
         # index-like
-        elif isinstance(data, (np.ndarray, Index, ABCSeries)):
+        elif isinstance(data, (np.ndarray, ABCIndexClass, ABCSeries)):
```
Comment (WillAyd) [inline pandas/core/indexes/base.py]: Can you limit scope of change in this module to just `ABCExtensionArray`? I think that's the only one that can be exclusively replaced with the ABC... import

The rest let's leave separate

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
lgtm can you merge master and try again? I think should resolve CI

**Comment by Kazz47** [comment]
Rebased, we'll see how this goes.

**Review by jreback — CHANGES_REQUESTED** [review]
this looks ok, can you run the asv suite (take a while) and see if this has any perf impact, you might want to start with a small subset (maybe the reshaping asvs)

**Comment by Kazz47** [comment]
@jreback When comparing 2ce2f4ff and 9f038370 there is no significant difference on the reshaping benchmark.

[reshaping-results.zip](https://github.com/pandas-dev/pandas/files/3748813/reshaping-results.zip)

**Comment by WillAyd** [comment]
@Kazz47 do you have bash output available you can post instead? Would be easier to review

**Comment by WillAyd** [comment]
@Kazz47 can you merge master and fix conflict? Also for asv run output just post the output from your terminal

**Comment by WillAyd** [comment]
Nice PR but needs rebase and I think has gone stale. Ping if you'd like to pick this back up
